File: tests/unit/test_ve_root_cause_diagnostics.py
"""Tests for Phase 7 Slice 7.7 — root-cause diagnostics surface (read-only).

Covers:
- Empty / under-threshold proposal sets produce no diagnostics
- Healthy data (no systemic patterns) produces no diagnostics
- Uniform global bias → injector_flow_error
- Low-load-only bias → deadtime_error
- Opposite high-vs-low load bias → target_table_error
- Correction strongly correlated with load axis → sensor_calibration_error
- Diagnostics never modify the input proposals
"""
from __future__ import annotations

from tuner.services.ve_analyze_cell_hit_service import (
    VeAnalysisProposal,
    VeAnalysisSummary,
)
from tuner.services.ve_root_cause_diagnostics_service import (
    VeRootCauseDiagnosticsService,
)


def _proposal(row: int, col: int, cf: float, n: int = 5) -> VeAnalysisProposal:
    return VeAnalysisProposal(
        row_index=row, col_index=col,
        current_ve=50.0, proposed_ve=round(50.0 * cf, 2),
        correction_factor=cf, sample_count=n,
    )


def _summary(*proposals: VeAnalysisProposal) -> VeAnalysisSummary:
    return VeAnalysisSummary(
        total_records=len(proposals),
        accepted_records=len(proposals),
        rejected_records=0,
        cells_with_data=len(proposals),
        cells_with_proposals=len(proposals),
        cell_corrections=(),
        proposals=proposals,
        rejection_counts_by_gate=(),
        summary_text="",
        detail_lines=(),
    )


def _rules(report) -> set[str]:
    return {d.rule for d in report.diagnostics}


# ---------------------------------------------------------------------------
# Threshold gating
# ---------------------------------------------------------------------------

class TestThresholdGating:
    def test_empty_summary_no_diagnostics(self) -> None:
        report = VeRootCauseDiagnosticsService().diagnose(_summary())
        assert report.diagnostics == ()
        assert report.has_findings is False

    def test_below_minimum_proposals_no_diagnostics(self) -> None:
        # 5 proposals — below the _MIN_PROPOSALS threshold of 6
        proposals = tuple(_proposal(0, c, 1.20) for c in range(5))
        report = VeRootCauseDiagnosticsService().diagnose(_summary(*proposals))
        assert report.diagnostics == ()


# ---------------------------------------------------------------------------
# Healthy data — no diagnostics fire
# ---------------------------------------------------------------------------

class TestHealthyData:
    def test_random_small_corrections_no_diagnostics(self) -> None:
        # 9 cells with corrections within ±2% of 1.0 — under all thresholds
        cfs = [0.98, 1.01, 0.99, 1.02, 0.99, 1.01, 1.00, 0.98, 1.02]
        proposals = tuple(
            _proposal(i // 3, i % 3, cf) for i, cf in enumerate(cfs)
        )
        report = VeRootCauseDiagnosticsService().diagnose(_summary(*proposals))
        assert report.diagnostics == ()


# ---------------------------------------------------------------------------
# injector_flow_error
# ---------------------------------------------------------------------------

class TestInjectorFlowError:
    def test_uniform_lean_bias_fires(self) -> None:
        # 9 cells all biased ~10% lean with low variance
        cfs = [1.10, 1.10, 1.10, 1.11, 1.10, 1.09, 1.10, 1.10, 1.10]
        proposals = tuple(
            _proposal(i // 3, i % 3, cf) for i, cf in enumerate(cfs)
        )
        report = VeRootCauseDiagnosticsService().diagnose(_summary(*proposals))
        assert "injector_flow_error" in _rules(report)
        finding = next(
            d for d in report.diagnostics if d.rule == "injector_flow_error"
        )
        assert "lean" in finding.message
        assert finding.severity == "warning"

    def test_uniform_rich_bias_fires(self) -> None:
        cfs = [0.90] * 9
        proposals = tuple(
            _proposal(i // 3, i % 3, cf) for i, cf in enumerate(cfs)
        )
        report = VeRootCauseDiagnosticsService().diagnose(_summary(*proposals))
        finding = next(
            d for d in report.diagnostics if d.rule == "injector_flow_error"
        )
        assert "rich" in finding.message

    def test_high_variance_does_not_fire(self) -> None:
        # Mean is biased lean but variance is too high
        cfs = [1.30, 0.90, 1.20, 0.95, 1.25, 0.88, 1.15, 0.92, 1.10]
        proposals = tuple(
            _proposal(i // 3, i % 3, cf) for i, cf in enumerate(cfs)
        )
        report = VeRootCauseDiagnosticsService().diagnose(_summary(*proposals))
        assert "injector_flow_error" not in _rules(report)


# ---------------------------------------------------------------------------
# deadtime_error
# ---------------------------------------------------------------------------

class TestDeadtimeError:
    def test_low_load_low_rpm_bias_fires(self) -> None:
        # 4×4 grid; low-load/low-rpm corner (rows 0-1, cols 0-1) lean by 15%;
        # rest near 1.0
        proposals: list[VeAnalysisProposal] = []
        for r in range(4):
            for c in range(4):
                if r <= 1 and c <= 1:
                    proposals.append(_proposal(r, c, 1.15))
                else:
                    proposals.append(_proposal(r, c, 1.00))
        report = VeRootCauseDiagnosticsService().diagnose(_summary(*proposals))
        assert "deadtime_error" in _rules(report)


# ---------------------------------------------------------------------------
# target_table_error
# ---------------------------------------------------------------------------

class TestTargetTableError:
    def test_opposite_high_low_load_bias_fires(self) -> None:
        # rows 0-1 lean (1.10), rows 2-3 rich (0.90), 4 cells per row
        proposals: list[VeAnalysisProposal] = []
        for r in range(4):
            cf = 1.10 if r <= 1 else 0.90
            for c in range(4):
                proposals.append(_proposal(r, c, cf))
        report = VeRootCauseDiagnosticsService().diagnose(_summary(*proposals))
        assert "target_table_error" in _rules(report)


# ---------------------------------------------------------------------------
# sensor_calibration_error
# ---------------------------------------------------------------------------

class TestSensorCalibrationError:
    def test_load_correlated_bias_fires(self) -> None:
        # Linear ramp of correction with row (load) — perfect correlation.
        # Use opposite-sign ramp so target_table_error doesn't dominate
        # the assertion (we just want to verify sensor_calibration fires).
        proposals: list[VeAnalysisProposal] = []
        for r in range(4):
            cf = 0.90 + r * 0.07  # 0.90, 0.97, 1.04, 1.11
            for c in range(3):
                proposals.append(_proposal(r, c, cf))
        report = VeRootCauseDiagnosticsService().diagnose(_summary(*proposals))
        assert "sensor_calibration_error" in _rules(report)
        finding = next(
            d for d in report.diagnostics if d.rule == "sensor_calibration_error"
        )
        assert "Pearson" in finding.message


# ---------------------------------------------------------------------------
# Read-only guarantee
# ---------------------------------------------------------------------------

class TestReadOnly:
    def test_diagnostics_do_not_mutate_input(self) -> None:
        cfs = [1.10] * 9
        proposals = tuple(
            _proposal(i // 3, i % 3, cf) for i, cf in enumerate(cfs)
        )
        summary = _summary(*proposals)
        before = tuple(
            (p.proposed_ve, p.correction_factor) for p in summary.proposals
        )
        VeRootCauseDiagnosticsService().diagnose(summary)
        after = tuple(
            (p.proposed_ve, p.correction_factor) for p in summary.proposals
        )
        assert before == after

File: cpp/tests/test_ve_root_cause_diagnostics.cpp
// SPDX-License-Identifier: MIT
//
// doctest unit tests for tuner_core::ve_root_cause_diagnostics.

#include "doctest.h"

#include "tuner_core/ve_root_cause_diagnostics.hpp"

namespace rcd = tuner_core::ve_root_cause_diagnostics;
namespace vps = tuner_core::ve_proposal_smoothing;

namespace {

vps::Proposal make(int r, int c, double cf) {
    vps::Proposal p;
    p.row_index = r;
    p.col_index = c;
    p.current_ve = 100.0;
    p.proposed_ve = 100.0 * cf;
    p.correction_factor = cf;
    p.sample_count = 1;
    return p;
}

}  // namespace

TEST_CASE("below MIN_PROPOSALS produces an empty report with explanatory text") {
    std::vector<vps::Proposal> ps;
    for (int i = 0; i < 3; ++i) ps.push_back(make(0, i, 1.10));
    auto r = rcd::diagnose(ps);
    CHECK(r.diagnostics.empty());
    CHECK_FALSE(r.has_findings());
    CHECK(r.summary_text.find("only 3 proposal") != std::string::npos);
}

TEST_CASE("healthy data produces no diagnostics") {
    std::vector<vps::Proposal> ps;
    // Mixed corrections, none big, no correlation.
    ps.push_back(make(0, 0, 1.01));
    ps.push_back(make(0, 1, 0.99));
    ps.push_back(make(1, 0, 1.00));
    ps.push_back(make(1, 1, 1.02));
    ps.push_back(make(2, 0, 0.98));
    ps.push_back(make(2, 1, 1.01));
    auto r = rcd::diagnose(ps);
    CHECK(r.diagnostics.empty());
    CHECK(r.summary_text.find("no systemic patterns") != std::string::npos);
}

TEST_CASE("uniform lean bias fires injector_flow_error") {
    std::vector<vps::Proposal> ps;
    for (int r = 0; r < 3; ++r)
        for (int c = 0; c < 3; ++c)
            ps.push_back(make(r, c, 1.10));  // 10% lean everywhere
    auto rep = rcd::diagnose(ps);
    bool found = false;
    for (const auto& d : rep.diagnostics) {
        if (d.rule == "injector_flow_error") {
            found = true;
            CHECK(d.severity == "warning");
            CHECK(d.message.find("lean") != std::string::npos);
        }
    }
    CHECK(found);
}

TEST_CASE("uniform rich bias fires injector_flow_error with rich label") {
    std::vector<vps::Proposal> ps;
    for (int r = 0; r < 3; ++r)
        for (int c = 0; c < 3; ++c)
            ps.push_back(make(r, c, 0.90));
    auto rep = rcd::diagnose(ps);
    bool found = false;
    for (const auto& d : rep.diagnostics) {
        if (d.rule == "injector_flow_error") {
            found = true;
            CHECK(d.message.find("rich") != std::string::npos);
        }
    }
    CHECK(found);
}

TEST_CASE("high variance suppresses injector_flow_error") {
    std::vector<vps::Proposal> ps;
    // Mean ~1.10 but huge variance — should not fire uniform-bias rule.
    ps.push_back(make(0, 0, 1.30));
    ps.push_back(make(0, 1, 0.90));
    ps.push_back(make(1, 0, 1.30));
    ps.push_back(make(1, 1, 0.90));
    ps.push_back(make(2, 0, 1.30));
    ps.push_back(make(2, 1, 0.90));
    auto rep = rcd::diagnose(ps);
    for (const auto& d : rep.diagnostics) {
        CHECK(d.rule != "injector_flow_error");
    }
}

TEST_CASE("deadtime error fires when low-load region is biased more than rest") {
    std::vector<vps::Proposal> ps;
    // 4x4 grid; bottom-left quadrant strongly lean, rest near unity.
    for (int r = 0; r < 4; ++r) {
        for (int c = 0; c < 4; ++c) {
            double cf = (r <= 1 && c <= 1) ? 1.20 : 1.00;
            ps.push_back(make(r, c, cf));
        }
    }
    auto rep = rcd::diagnose(ps);
    bool found = false;
    for (const auto& d : rep.diagnostics) {
        if (d.rule == "deadtime_error") {
            found = true;
            CHECK(d.severity == "warning");
        }
    }
    CHECK(found);
}

TEST_CASE("opposite high/low load bias fires target_table_error") {
    std::vector<vps::Proposal> ps;
    // 4 rows; low rows lean (cf>1), high rows rich (cf<1).
    for (int c = 0; c < 3; ++c) {
        ps.push_back(make(0, c, 1.08));
        ps.push_back(make(1, c, 1.07));
        ps.push_back(make(2, c, 0.93));
        ps.push_back(make(3, c, 0.92));
    }
    auto rep = rcd::diagnose(ps);
    bool found = false;
    for (const auto& d : rep.diagnostics) {
        if (d.rule == "target_table_error") {
            found = true;
            CHECK(d.severity == "info");
        }
    }
    CHECK(found);
}

TEST_CASE("strong load-axis correlation fires sensor_calibration_error") {
    std::vector<vps::Proposal> ps;
    // Linear ramp: row index 0..5 → cf 1.00..1.25
    for (int r = 0; r < 6; ++r) {
        ps.push_back(make(r, 0, 1.00 + 0.05 * r));
    }
    auto rep = rcd::diagnose(ps);
    bool found = false;
    for (const auto& d : rep.diagnostics) {
        if (d.rule == "sensor_calibration_error") {
            found = true;
            CHECK(d.severity == "info");
            CHECK(d.message.find("Pearson r=") != std::string::npos);
        }
    }
    CHECK(found);
}

TEST_CASE("input is never mutated") {
    std::vector<vps::Proposal> ps;
    for (int r = 0; r < 3; ++r)
        for (int c = 0; c < 3; ++c)
            ps.push_back(make(r, c, 1.10));
    rcd::diagnose(ps);
    CHECK(ps.size() == 9);
    CHECK(ps[0].correction_factor == doctest::Approx(1.10));
}

TEST_CASE("summary text format pin for findings") {
    std::vector<vps::Proposal> ps;
    for (int r = 0; r < 3; ++r)
        for (int c = 0; c < 3; ++c)
            ps.push_back(make(r, c, 1.10));
    auto rep = rcd::diagnose(ps);
    CHECK(rep.has_findings());
    CHECK(rep.summary_text.find("Root-cause diagnostics:") == 0);
    CHECK(rep.summary_text.find("pattern(s) found") != std::string::npos);
}
